packages/clay-modal/src/Provider.tsx

```diff
@@ -0,0 +1,91 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import ClayModal from './Modal';
+import React, {createContext, useReducer} from 'react';
+import {Size, Status} from './types';
+
+enum Action {
+	Close = 0,
+	Open = 1,
+}
+
+interface IProps {
+	spritemap?: string;
+}
+
+type TState = {
+	body: React.ReactElement | React.ReactText;
+	footerFirst?: React.ReactElement;
+	footerLast?: React.ReactElement;
+	header?: React.ReactElement | React.ReactText;
+	size?: Size;
+	status?: Status;
+	url?: string;
+};
+
+type TAction = {type: Action.Open; payload: TState} | {type: Action.Close};
+
+type TProvider = [TState, React.Dispatch<TAction>];
+
+const initialState = {
+	body: <div />,
+	visible: false,
```

Comment:
Not sure if I am fully understanding this right. But would this not allow for multiple modals? We might be able to leverage a Map for this and have unique ids for each modal open.

Reply:
It won't create multiple modals, we just want to have one Modal open, so calling a dispatch will change the existing context information. Because you think you would create multiple modals?

Reply:
I am primarily thinking of a context where a developer might need nested models, they open one and then a button opens another one on top of that. I can imagine devs using it this way, even if its not necessarily intended, so I think we might want to think of having a global Map of open Models. That way we can close a specific modal, instead of closing everything at once.

Reply:
Actually I wanted to discourage the use of nested modals is something we should not encourage, this is not correct.

Reply:
I think we can still discourage it, but not necessarily outlaw it. Perhaps a normal use case would be a modal open and then an additional "Confirm" type modal/dialog on top.

It may not be needed quite yet, but perhaps we can add this down the line if its too much work to add in this PR

Reply:
I say this because I had a conversation implementing this PR about it https://github.com/liferay/clay/pull/1842#discussion_r278464379 and I also think calling modal within another modal is a wrong way to do it, Lexicon also seems not to recommend https://liferay.design/lexicon/core-components/modals.

Reply:
As far as I know from every conversation I've had with Lexicon over the past 2 years, modals over modals are an anti-pattern to be avoided at all costs 😂 

Ideally, we should find some way to navigate within a modal back and forth rather than stacking them. This is mostly a UX issue that we can't really tackle at once because it means changing a lot of workflows and components, so we've always had to just take it as it is.